crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2178 @@
+     * EVP_EncryptUpdate can be called multiple times if necessary
+     */
+    if (EVP_EncryptUpdate(ctx, ciphertext, &len, plain, plainlen) != 1) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    ciphertextlen = len;
+    /* Finalise the encryption. */
+    if (EVP_EncryptFinal_ex(ctx, ciphertext + len, &len) != 1) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    ciphertextlen += len;
+    /*
+     * Get the tag This isn't a duplicate so needs to be added to the ciphertext
+     */
+    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, taglen, tag) != 1) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    memcpy(ciphertext + ciphertextlen, tag, taglen);
+    ciphertextlen += taglen;
+    if (ciphertextlen > *cipherlen) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    *cipherlen = ciphertextlen;
+    memcpy(cipher, ciphertext, ciphertextlen);
+
+err:
+    EVP_CIPHER_CTX_free(ctx);
+    EVP_CIPHER_free(enc);
+    OPENSSL_free(ciphertext);
+    return erv;
+}
+
+/*
+ * @brief check mode is in-range and supported
+ * @param mode is the caller's chosen mode
+ * @return 1 for good (OpenSSL style), not 1 for error
+ */
+static int hpke_mode_check(unsigned int mode)
+{
+    switch (mode) {
+    case OSSL_HPKE_MODE_BASE:
+    case OSSL_HPKE_MODE_PSK:
+    case OSSL_HPKE_MODE_AUTH:
+    case OSSL_HPKE_MODE_PSKAUTH:
+        break;
+    default:
+        return 0;
+    }
+    return 1;
+}
+/**
+ * @brief check if a suite is supported locally
+ *
+ * @param suite is the suite to check
+ * @return 1 for good/supported, not 1 otherwise
+ */
+static int hpke_suite_check(OSSL_HPKE_SUITE suite)
+{
+    /*
+     * Check that the fields of the suite are each
+     * implemented here, loops below start at one
+     * as zero'th element of tables is "null"
+     */
+    int kem_ok = 0;
+    int kdf_ok = 0;
+    int aead_ok = 0;
+    int ind = 0;
+    int nkems = OSSL_NELEM(hpke_kem_tab);
+    int nkdfs = OSSL_NELEM(hpke_kdf_tab);
+    int naeads = OSSL_NELEM(hpke_aead_tab);
+
+    /* check KEM */
+    for (ind = 1; ind != nkems; ind++) {
+        if (suite.kem_id == hpke_kem_tab[ind].kem_id) {
+            kem_ok = 1;
+            break;
+        }
+    }
+
+    /* check kdf */
+    for (ind = 1; ind != nkdfs; ind++) {
+        if (suite.kdf_id == hpke_kdf_tab[ind].kdf_id) {
+            kdf_ok = 1;
+            break;
+        }
+    }
+
+    /* check aead */
+    for (ind = 1; ind != naeads; ind++) {
+        if (suite.aead_id == hpke_aead_tab[ind].aead_id) {
+            aead_ok = 1;
+            break;
+        }
+    }
+
+    if (kem_ok == 1 && kdf_ok == 1 && aead_ok == 1)
+        return 1;
+    return 0;
+}
+
+/*
+ * @brief generate a key pair keeping private inside API
```

Comment:
The function returns the private - not sure what this comment means

Reply:
legacy stuff - there used be an API that returned the private key in a buffer too
fixed the comment